File: Data Structures & Algorithms/course-schedule-ii/submission-6.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        preq = {i:[] for i in range(numCourses)}
        for c, p in prerequisites:
            preq[c].append(p)
        ans = []
        visited = [0 for _ in range(numCourses)]
        def dfs(i):
            if visited[i] == 2:
                return True
            if visited[i] == 1:
                return False
            visited[i] = 1
            for nei in preq[i]:
                if not dfs(nei):
                    return False
            ans.append(i)
            visited[i] = 2
            return True
        for i in range(numCourses):
            if not dfs(i):
                return []
        return ans        

File: Data Structures & Algorithms/course-schedule/submission-6.cpp
#include <iostream>
#include <vector>
#include <numeric> // Required for std::accumulate

class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<int> indegree(numCourses, 0);
        vector<vector<int>> graph(numCourses, vector<int>());
        vector<int> result;
        for (auto& p : prerequisites) {
            indegree[p[1]]++;
            graph[p[0]].push_back(p[1]);
        }
        queue<int> que;
        for (int i = 0; i < numCourses; i++) {
            if (indegree[i] == 0) {
                que.push(i);
            }
        }
        while (!que.empty()) {
            int curr = que.front();
            que.pop();
            for (const int nei : graph[curr]) {
                indegree[nei]--;
                if (indegree[nei] == 0) {
                    que.push(nei);
                }
            }
        }

        return accumulate(indegree.begin(), indegree.end(), 0) == 0;
    }
};
